Signal/src/Models/AccountManager.swift

```diff
@@ -33,24 +34,33 @@ class AccountManager: NSObject {
     }
 
     func register(verificationCode: String) -> Promise<Void> {
-        let registrationPromise = firstly {
-            Promise { fulfill, reject in
-                if verificationCode.characters.count == 0 {
-                    let error = OWSErrorWithCodeDescription(.userError,
-                                                            NSLocalizedString("REGISTRATION_ERROR_BLANK_VERIFICATION_CODE",
-                                                                              comment: "alert body during registration"))
-                    reject(error)
-                }
-                fulfill()
-            }
-        }.then {
-            Logger.debug("\(self.TAG) verification code looks well formed.")
-            return self.registerForTextSecure(verificationCode: verificationCode)
+        guard verificationCode.characters.count > 0 else {
+            let error = OWSErrorWithCodeDescription(.userError,
+                                                    NSLocalizedString("REGISTRATION_ERROR_BLANK_VERIFICATION_CODE",
+                                                                      comment: "alert body during registration"))
+            return Promise(error: error)
+        }
+
+        Logger.debug("\(self.TAG) registering with signal server")
+        let registrationPromise: Promise<Void> = firstly {
+            self.registerForTextSecure(verificationCode: verificationCode)
         }.then {
-            return SyncPushTokensJob.run(pushManager: self.pushManager, accountManager: self, preferences: self.preferences)
+            self.syncPushTokens()
+        }.recover { (error) -> Promise<Void> in
+            switch error {
+            case PushRegistrationError.pushNotSupported(let description):
+                // This can happen with:
+                // - simulators, none of which support receiving push notifications
+                // - on iOS11 devices which have disabled "Allow Notifications" and disabled "Enable Background Refresh" in the system settings.
+                Logger.info("\(self.TAG) Recovered push registration error. Registering for manual message fetcher because push not supported: \(description)")
+                return self.registerForManualMessageFetching()
```

Comment:
To receive messages, a device must either:

1. register push tokens (this is the only way we used to work)
2. specify that they'll fetch messages manually (this is how Desktop always works)

Note:  for simulators we no longer register fake push tokens.

Reply:
If user later gets a push token, do we need to "unregister for manual message fetching"?

Reply:
I asked moxie the same thing and he said "no need to do that."